app/src/org/commcare/android/framework/ManagedUiFramework.java

```diff
@@ -122,15 +135,22 @@ private static boolean isFieldInBundle(String elementKey, Bundle bundle) {
      * Store the state of the activity's UiElement annotated fields.
      */
     public static Bundle saveUiStateToBundle(CommCareActivity activity) {
+
         Bundle bundle = new Bundle();
-        Class c = activity.getClass();
-        for (Field f : c.getDeclaredFields()) {
+        Class classHoldingFields = getClassHoldingFields(activity);
+
+        for (Field f : classHoldingFields.getDeclaredFields()) {
             if (f.isAnnotationPresent(UiElement.class)) {
                 UiElement element = f.getAnnotation(UiElement.class);
                 try {
                     f.setAccessible(true);
                     try {
-                        View v = (View)f.get(activity);
+                        View v;
+                        if (activity.usesUIController()) {
+                            v = (View)f.get(activity.getUIController());
+                        } else {
+                            v = (View)f.get(activity);
+                        }
```

Comment:
I would lift this out of the for loop, such that you get the proper class instance once. It also sort of  coincides with the result of `getClassHoldingFields`, so maybe some of the setup work can be shared.